src/bin/pg_autoctl/cli_config.c

```diff
@@ -632,6 +632,157 @@ cli_config_set(int argc, char **argv)
 }
 
 
+static bool
+cli_keeper_config_validate_replication(KeeperConfig *config, PGSQL *pgsql)
+{
+	Keeper keeper = { 0 };
+	NodeAddress primary = { 0 };
+	bool primaryUnreachable;
+
+	if (!keeper_init(&keeper, config))
+	{
+		/* already logged */
+		return false;
+	}
+
+	/* get the primary info to populate the replication source */
+	if (!monitor_get_primary(&keeper.monitor,
+							 config->formation,
+							 keeper.state.current_group,
+							 &primary))
+	{
+		log_warn("Monitor could not be contacted for primary conninfo, skipping "
+				 "password validation");
+		return true;
+	}
+
+	if (primary.nodeId == keeper.state.current_node_id)
+	{
+		/* we are the primary */
+		PGSQL *pgsql = &keeper.postgres.sqlClient;
+
+		/* Skip validation if the primary isn't running. */
+		if (!pg_is_running(config->pgSetup.pg_ctl, config->pgSetup.pgdata))
+		{
+			log_warn("Primary could not be contacted, skipping password "
+					 "validation");
+			return true;
+		}
+
+		/*
+		 * Validate the replication password in a transaction block by setting
+		 * the password using ALTER USER PASSWORD. The transaction block will
+		 * encapsulate the subsequent change to the config file as well and we
+		 * ROLLBACK/COMMIT depending on whether both the ALTER and the config
+		 * file update are successful.
+		 */
+		if (!pgsql_execute(pgsql, "BEGIN"))
+		{
+			log_error("Failed to open a SQL transaction to update the "
+					  "replication password");
+			pgsql_finish(pgsql);
+			return false;
+		}
+
+		if (!pgsql_set_password(pgsql,
+								PG_AUTOCTL_REPLICA_USERNAME,
+								config->replication_password))
+		{
+			log_error("Failed to set replication password, see above for details");
+			pgsql_finish(pgsql);
+			return false;
+		}
+
+		/*
+		 * Copy the SQL connection state to the caller, for later
+		 * COMMIT/ROLLBACK.
+		 */
+		*pgsql = keeper.postgres.sqlClient;
```

Comment:
I think this is a big no: having code that handles an SQL transaction from multiple functions, each doing early `return false`, without proper documentation about it, and when it's very unclear that this level of complexity is needed at all... please let's try another approach.

I think this whole code be way easier to read/write/maintain when done the following way:
  - check what setting is being set from the command line, and when `replication.password` invoke a specific function,
  - the `pg_autoctl set replication.password` specific function either forwards the change to Postgres, on the primary, or change the `primary_conninfo` connection string, on a secondary ; again, I think this first function should only recognise the current situation and then delegate to a specialised function,
  - now we have a function that has to ALTER ROLE on the primary and that's it,
  - and another function that has to use `pgctl_identify_system` to validate the new connection string, and when successful (maybe after some retries), then restart the secondary with the new replication settings.

Your implementation seems to be more complex than that in order to enable a smart handling of keeping the success of the transaction (ALTER ROLE on the primary, establishing a replication connection on the secondary) and the actual editing of the configuration file. What's complex is to be able to ROLLBACK the ALTER ROLE command when the configuration file edit failed. For that, I think it's fair to say that the main CLI entry point should delegate the configuration file handling to the `pg_autoctl set replication.password` specific function(s), and that the BEGIN/COMMIT/ROLLBACK and file update is only necessary in the sub-function that takes care of the primary. There's no Postgres transaction in the control flow of the secondary side of things.

Reply:
w/ @jchampio

> I think this whole code be way easier to read/write/maintain when done the following way:

We did try that approach _(We can try the approach again and present what it looks like)._ However we had a few pain points:

* There was a lot of code duplication (a lot of the keeper setup code and `keeper_config_write_file`) between the function to handle the primary case and the function to handle the secondary case.

* In addition to the code duplication, if we want to validate the entire config holistically (one config value change might affect other config values adversely as @jchampio mentions [here](https://github.com/citusdata/pg_auto_failover/pull/399#discussion_r513629674)), then we can't have a specialized function.

> and another function that has to use pgctl_identify_system to validate the new connection string, and when successful (maybe after some retries), then restart the secondary with the new replication settings.

The subsequent config reload (`cli_pg_autoctl_reload()` call) would lead to `keeper_ensure_configuration()` restarting the standby anyway. Do we want the replication settings update in the recovery conf file and an explicit restart from the client? Is this to do with ensuring that the effect of a `pg_autoctl config set` command is immediately enforced?

Reply:
I think some code duplication is much less of a problem than complex transaction management in different independent functions with early exists for error handling. Let's make it right, and if we can get rid of code duplication in a simple way, we can do it later. Now it's impossible for me to follow all the possible cases where we might exit with a transaction left open.

About configuration settings that have an effect on the primary_conninfo, please list them. I still don't think it would be wise to recompute and test our primary_conninfo when the user changes our `network_partition_timeout` for instance.

Then about restarting Postgres, you're right, that sounds like a job for the background process upon processing a RELOAD signal.